docs/styleguide/PyGuide.md

```diff
@@ -816,3 +820,74 @@ This is required so that the autogenerated API documentation is rendered properl
+<a id="6-test-suite-coding-rules"></a>
+<a id="test-suite-coding-rules"></a>
+
+## 6 Test suite coding rules
+
+Unit tests are written and executed using the [pytest](https://pytest.org) framework.
+
+Parameters of test cases and fixtures should be type-hinted just like any other function.
+Test cases should ideally all have docstrings, except for the simple ones which are sufficiently described either by their names or by the simplicity of the code.
+
+For purposes of setting up test environment for a given case, or reusing setup/teardown code, or parametrizing complex cases the [fixture](https://docs.pytest.org/en/7.4.x/explanation/fixtures.html#about-fixtures) approach should be used.
+[Fixture scoping](https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session) should be leveraged depending on the use case - for instance, a class-scoped fixture can be utilized to provide a storage object for data gathered in individual cases inside the class, and to post-process this data after the test cases in the class have all been executed.
+The fixtures that are meant to be reused in multiple test files should be defined in a corresponding `conftest.py` file, instead of importing these where they are needed from another non-`conftest.py` file where they were defined - otherwise the linters will try to remove it as an unused import and also the import itself may have an effect on how the fixture is executed based on its scope.
+
+Test cases may be parametrized either using a `@pytest.mark.parametrize` decorator (for simple parametrizations), or by using [parametrized fixtures](https://docs.pytest.org/en/6.2.x/fixture.html#parametrizing-fixtures) in case the test requires complex parametrization, or in case the parameters have to be preprocessed in a complex fashion before being passed into the test.
+The parameter set ("test structs") for each test case should be represented as a collection of type-hinted class objects (or dataclass objects).
+Do not use tuples, or namedtuples, or dicts to represent an individual parameter - these cannot be typed or refactored effectively, and for complex cases it is hard to visually tell by looking at the definiton of the parameter (test struct) just what each sub-parameter means.
+When instantiating a test struct object, specify init arguments explicitly as keywords for increased visibility.
+
+Bad:
+
+```python3
+@pytest.mark.parametrize([(42, "foo", None),
+                          (1337, "bar", RuntimeError)])
+def test_case_parametrized(param):
+    assert function_call() == param[0]
+    assert another_object.attribute != param[1]
+    if param[2] is not None:
+      with pytest.raises(param[2]):
+        should_raise()
+```
+
+Good:
+
+```python3
+@dataclass
+class StructForTest:
+  expected_return_value: int
+  prohibited_attribute_value: str
+  raises: Optional[Type[Exception]]
+  
+@pytest.mark.parametrize([StructForTest(expected_return_value=42,
```

Comment:
Missed name of parameter
```python 
@pytest.mark.parametrize(
    "ts",  # <----
    [
        StructForTest(expected_return_value=42, prohibited_attribute_value="foo", raises=None),
        StructForTest(expected_return_value=1337, prohibited_attribute_value="bar", raises=RuntimeError),
    ],
)
```

Please provide code after black formatting.
And looks like need `idfn` for this case.

Reply:
Done